synapse/http/server.py

```diff
@@ -499,6 +500,71 @@ class RootOptionsRedirectResource(OptionsResource, RootRedirect):
     pass
 
 
+@implementer(interfaces.IPullProducer)
+class _ByteProducer:
+    """
+    Iteratively write bytes to the request.
+    """
+
+    # The minimum number of bytes for each chunk. Note that the last chunk will
+    # usually be smaller than this.
+    min_chunk_size = 1024
+
+    def __init__(
+        self, request: Request, iterator: Iterator[bytes],
+    ):
+        self._request = request
+        self._iterator = iterator
+
+    def start(self):
+        self._request.registerProducer(self, False)
+
+    def _send_data(self, data: List[bytes]) -> None:
+        """
+        Send a list of strings as a response to the request.
+        """
+        if not data:
+            return
+        self._request.write(b"".join(data))
+
+    def resumeProducing(self):
+        # We've stopped producing in the meantime.
+        if not self._request:
+            return
+
+        # Get the next chunk and write it to the request. Calling write will
+        # spin the reactor (and might be re-entrant).
+        buffer = []
+        buffered_bytes = 0
+        while buffered_bytes < self.min_chunk_size:
+            try:
+                data = next(self._iterator)
+                buffer.append(data)
+                buffered_bytes += len(data)
```

Comment:
Any reason not to use `len(buffer)`? I imagine that is just as fast

Reply:
The buffer is a list of chunks, so `len(buffer)` is the number of chunks, *not* the number of bytes.

The goal of this was to avoid continually expanding a string which is usually not great for speed.

Reply:
Aha, I misread the code :)

Reply:
I added a bit of the above reasoning to the comment above `buffer`.